crates/typst-library/src/visualize/gradient.rs

```diff
@@ -582,12 +582,11 @@ impl Gradient {
                 let mut stops = stops
                     .iter()
                     .map(move |&(color, offset)| {
-                        let t = i as f64 / n as f64;
                         let r = offset.get();
                         if i % 2 == 1 && mirror {
-                            (color, Ratio::new(t + (1.0 - r) / n as f64))
+                            (color, Ratio::new((i as f64 + 1.0 - r) / n as f64))
                         } else {
-                            (color, Ratio::new(t + r / n as f64))
+                            (color, Ratio::new((i as f64 + r) / n as f64))
```

Comment:
I have a feeling that this doesn't fully solve the problem, since there is still some potential for imprecision. But it's nice that this improves the situation nonetheless by reducing the amount of divisions. Perhaps using fixed point numbers (a future task, not suitable for this PR) would be required to fully fix the issue.

I wonder if doing something like

```rs
Ratio::new(((i + 1) as f64 - r) / n as f64)
```

on the first one could improve it even further by relying more on integer arithmetic.

Reply:
It would be a bit cleaner, but I think it wouldn't make a material difference as f64 can accurately represent ints even beyond 2^32 and the indices won't be that large.

Reply:
Okay